[LLDB] Impove ObjectFileELF's .dynamic parsing and usage.

clayborg opened the pull request:
This patch improves the ability of a ObjectFileELF instance to read the .dynamic section. It adds the ability to read the .dynamic section from a ELF file that is read from memory, cleans up the usage of the .dynamic entries so that ObjectFileELF::ParseDynamicSymbols() is the only code that parses .dynamic entries, teaches that function the read and store the string values for each .dynamic entry, and dumps the .dynamic entries in the output of "image dump objfile". It also cleans up the code that gets the dynamic string table and the .dynamic data from the ELF file.

clayborg committed:
Impove ObjectFileELF's .dynamic parsing and usage.

This patch improves the ability of a ObjectFileELF instance to read the .dynamic section. It adds the ability to read the .dynamic section from a ELF file that is read from memory, cleans up the usage of the .dynamic entries so that ObjectFileELF::ParseDynamicSymbols() is the only code that parses .dynamic entries, teaches that function the read and store the string values for each .dynamic entry, and dumps the .dynamic entries in the output of "image dump objfile". It also cleans up the code that gets the dynamic string table and the .dynamic data from the ELF file.

clayborg committed:
Ran git-clang-format to fix clang format issues.

labath commented on lldb/test/Shell/ObjectFile/ELF/elf-memory.test:
```diff
@@ -0,0 +1,58 @@
+// REQUIRES: system-linux, native
```
I would like to see a test which uses an on-disk section-header-free object file. (Maybe you could base it off of llvm/test/tools/llvm-readobj/ELF/dynamic-reloc-no-section-headers.test). The in-memory test has a lot of moving parts, it seems pretty fragile(*), and can only run on linux (or at least an elf system). OTOH, an on-disk test would be very self contained and can run anywhere. If this breaks (which I'm sure it eventually will), I'd much rather be debugging the second one.

I know that in-memory object files are your main use case, and we should have a test for that, but I think it should be like a cherry on top rather than the first&only line of defense. 

(*) It relies on certain entries being produces by the linker. Most of these are pretty much guaranteed to be there, but for example DT_NEEDED(libc) may not exist if the c library is linked statically. If we can test all of the small pieces with files, then I think it'd be best if the overall in memory test was a full end-to-end API test, which actually deletes the object file, and then verifies that we're still able to read certain things from the object.

clayborg replied:
Adding support for memory ELF files is my primary concern with this patch as you know. Testing that in memory ELF files work as well as they can and that we can find the DT_DEBUG is the primary goal of this patch, so I would still like to concentrate on that. I have yet to run into an ELF file in the wild that doesn't have section headers, though I know they are possible. Testing I can load a ELF file from disk with no section headers is a whole other code path that introduces changes and time to this patch which I would rather not take on at this point if possible. Getting core files to be able to have a shared library list without having the main executable makes ELF core file actually useful now instead of being useless unless you get the correct executable. So although testing an on disk ELF file with no section headers is nice for testing, it can't replace the test I have written because we need this to test the in memory ELF files and testing the file itself is whole other code path that isn't needed.

I will be happy to add support for on disk ELF files after this patch, but I would like to get the useful part of this patch checked in so we can get ELF core files to be useful with no executable.

labath replied:
If it was a lot of work then sure, I'd agree with you -- but I don't think that's the case here. I just gave it a shot, and this is the only change I needed to make on top of your patch to be able to parse DT_NEEDED of a section-free file (disclaimer: This code is mainly for demonstration, it's possible there is a better way to do that):
```
diff --git a/lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp b/lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp
index d26215a9ce3f..5f0d2233b865 100644
--- a/lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp
+++ b/lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp
@@ -3810,28 +3810,26 @@ std::optional<DataExtractor> ObjectFileELF::GetDynstrData() {
     }
   }
 
-  // Every ELF file which represents an executable or shared library has
-  // mandatory .dynamic entries. Two of these values are DT_STRTAB and DT_STRSZ
-  // and represent the dynamic symbol tables's string table. These are needed
-  // by the dynamic loader and we can read them from a process' address space.
-  //
-  // When loading and ELF file from memory, only the program headers end up
-  // being mapped into memory, and we can find these values in the PT_DYNAMIC
-  // segment.
-  if (!IsInMemory())
-    return std::nullopt;
-  ProcessSP process_sp(m_process_wp.lock());
-  if (!process_sp)
-    return std::nullopt;
-
   const ELFDynamic *strtab = FindDynamicSymbol(DT_STRTAB);
   const ELFDynamic *strsz = FindDynamicSymbol(DT_STRSZ);
   if (strtab == nullptr || strsz == nullptr)
     return std::nullopt;
 
-  if (DataBufferSP data_sp =
-          ReadMemory(process_sp, strtab->d_ptr, strsz->d_val))
-    return DataExtractor(data_sp, GetByteOrder(), GetAddressByteSize());
+  if (ProcessSP process_sp = m_process_wp.lock()) {
+    if (DataBufferSP data_sp =
+            ReadMemory(process_sp, strtab->d_ptr, strsz->d_val))
+      return DataExtractor(data_sp, GetByteOrder(), GetAddressByteSize());
+  } else {
+    Address addr;
+    if (addr.ResolveAddressUsingFileSections(strtab->d_ptr, GetSectionList())) {
+      DataExtractor data;
+      addr.GetSection()->GetSectionData(data);
+      return DataExtractor(data,
+                           strtab->d_ptr - addr.GetSection()->GetFileOffset(),
+                           strsz->d_val);
+    }
+  }
+
   return std::nullopt;
 }
 
```

And I think this goes a long way towards improving the testability of this patch. I know section-free files aren't likely to be of much use, but they're really great for testing, and in terms of raw LOC, such a test may cover 80% of this patch.

clayborg replied:
I added a test with an on disk file

bulbazord commented on lldb/test/Shell/ObjectFile/ELF/elf-memory.test:
```diff
@@ -0,0 +1,58 @@
+// This test will make a simple executable that:
+//   - links against libc
+//   - runs and stops at a breakpoint
+//   - 
```
nit: Extra bullet point

bulbazord commented on lldb/test/Shell/ObjectFile/ELF/elf-memory.test:
```diff
@@ -0,0 +1,58 @@
+// REQUIRES: system-linux, native
+
+// This test verifies that loading and ELF file from memory works and the new
```
`loading and ELF file` -> `loading an ELF file`

bulbazord commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.h:
```diff
@@ -384,6 +392,9 @@ class ObjectFileELF : public lldb_private::ObjectFile {
   void DumpDependentModules(lldb_private::Stream *s);
 
+  /// ELF dump the .dynamic section
+  void DumpELFDynamic(lldb_private::Stream *s);
```
Can this take a `Stream &` instead of a `Stream *`? The implementation assumes that the pointer is valid, a reference would give us a better guarantee at compile time.

clayborg replied:
All other dump calls use a "Stream *" so I mirrored those. I can make a separate patch to switch these over to "Stream &". Would that be ok? I didn't want to make the patch bigger

labath replied:
I wouldn't say its required, but I'd like changing these to a reference (in another patch). If the patch is trivial/obvious, i don't think it requires a review.

bulbazord reviewed (commented):
Can't say I know a ton about ELF or your goals, but I found a few things I think could make the patch better.

I noticed you have a lot of this pattern:
```
Type *thingy = GetThingy();
if (thingy) {
```
I would suggest condensing that to:
```
if (Type *thingy = GetThingy()) {
```
That way the code that checks the validity of the pointer is the same code that defines the pointer, creating less opportunity for mistakes now and in the future.

clayborg commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.h:
```diff
@@ -402,6 +413,29 @@ class ObjectFileELF : public lldb_private::ObjectFile {
+  ///
+  /// \return The bytes that represent the string table data or \c std::nullopt
+  ///         if an error occured.
+  std::optional<lldb_private::DataExtractor> GetDynamicData();
```
Right now we just don't return anything if there is an error. I don't think returning an empty data makes anything better, as we would try to decode stuff from the DataExtractor and get all zeroes back and thinkg it was a valid DT_NULL entry. I can change this if you feel stronglym or switch to a `llvm::Expected<lldb_private::DataExtractor>` if that makes things better, though I am not sure anyone will care to see the error, but I guess we can log it if we have logging enabled otherwise consume the error. Let me know what you think.

labath replied:
I don't feel strongly, but my thinking is like:
- we should handle both the case where the dynamic section is missing (for whatever reason) and when it is present, but empty (for whatever reason)
- we probably want to handle both cases the same way (probably by doing nothing and bailing out)
- if we coalesce the two cases into one, the caller is more likely to get both of them right

clayborg commented on lldb/source/Target/Process.cpp:
```diff
@@ -2550,6 +2550,21 @@ ModuleSP Process::ReadModuleFromMemory(const FileSpec &file_spec,
+      size_to_read = 0x2000;
+      MemoryRegionInfo range_info;
+      Status error(GetMemoryRegionInfo(header_addr, range_info));
+      if (error.Success()) {
+        // We found a memory region, set the range of bytes ro read to read to
+        // the end of the memory region. This should be enough to contain the
+        // file header and important bits.
+        const auto &range = range_info.GetRange();
```
We don't care about the code, we care about getting all of the program headers. Without this change, my test doesn't work and the test verifies that this code is working. Prior to this, this was defaulting to 512 bytes, so we would sometimes get data back for the program headers if they started within the first 512 bytes, and sometimes we would get part of the program headers and they would be cut short because we didn't have all of the data. Other times we would get no bytes for the program headers. By getting the first memory segment for a binary, this fixed all 3 cases as it at least ensures we have enough info for the program headers. Once we have all program headers, we can now end up reading anything else we need from process memory, so any data pointed to by the program headers can be fetched from the process memory. I can remove this change and do the extra work in the ObjectFileELF.cpp when we load from memory, but I figured this would be useful for other things like mach-o and COFF too.

So in your example above, check to see if all program headers are in the first LOAD segment, if that is true, then this patch will work just fine on those.

labath replied:
They are, but that's not my point. My point is that this first segment can be really large (like one gigabyte or even more), and if that's the case, we probably don't want to read all of it.

Doing the extra read in ObjectFileELF sounds good to me, as there we can look at the elf header, and load exactly only the program header bytes we care about.

clayborg commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:
```diff
@@ -873,33 +873,29 @@ Address ObjectFileELF::GetImageInfoAddress(Target *target) {
+        return Address();
+
+      addr_t offset = (i * 2 + 1) * GetAddressByteSize();
       addr_t dyn_base = dynsym_section_sp->GetLoadBaseAddress(target);
       if (dyn_base == LLDB_INVALID_ADDRESS)
         return Address();
```
No, it requires immediate load address resolving and then it uses the address for something else. I have no hope of reproducing this so I left the code as is because I won't be able to test any changes for MIPS.

labath replied:
I'm not sure what you mean, the code does go about it in slightly different way (because that was convenient at the time), but I think it should be equivalent to:
```
addr_t offset = (i * 2 + 1) * GetAddressByteSize();
addr_t file_addr = m_dynamic_base_addr + offset;
Address addr;
if (!addr.ResolveAddressUsingFileSections(file_addr, GetSectionList()))
  return Address();
Status error;
if (symbol.d_tag == DT_MIPS_RLD_MAP) {
    // DT_MIPS_RLD_MAP tag stores an absolute address of the debug pointer.
    Address pointer_addr;
    if (target->ReadPointerFromMemory(addr, error, pointer_addr, true))
      return pointer_addr;
}
```
Basically, I think it does the same thing as DT_DEBUG, but with an extra level of indirection.


Also, this is kind of the reason why I'd really like to have a test for this patch which does not require running an executable. I don't want the next person to come in here to be afraid of making any changes because he can't test his change.

labath replied:
FWIW, I think noone actually uses this code. We definitely don't have any mips buildbots, so if anything breaks, it's up to them to fix it.

clayborg commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:
```diff
@@ -873,33 +873,29 @@ Address ObjectFileELF::GetImageInfoAddress(Target *target) {
     if (symbol.d_tag == DT_DEBUG) {
       // Compute the offset as the number of previous entries plus the size of
       // d_tag.
-      addr_t offset = i * dynsym_hdr->sh_entsize + GetAddressByteSize();
-      return Address(dynsym_section_sp, offset);
+      addr_t offset = (i * 2 + 1) * GetAddressByteSize();
+      addr_t file_addr = m_dynamic_base_addr + offset;
+      Address addr;
```
we can, but it will normally only happen for `DT_DEBUG` so I don't think we need this value for any other DT value.

labath replied:
This is mainly useful in conjuction with the mips change below.

clayborg commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:
```diff
@@ -3664,7 +3730,27 @@ llvm::ArrayRef<ELFProgramHeader> ObjectFileELF::ProgramHeaders() {
+    ProcessSP process_sp(m_process_wp.lock());
+    if (process_sp) {
+      const lldb::offset_t base_file_addr = GetBaseAddress().GetFileAddress();
+      // const addr_t data_addr = m_memory_addr + H.p_offset; // Not correct for
```
will do

clayborg commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:
```diff
@@ -3704,3 +3790,83 @@ ObjectFileELF::MapFileDataWritable(const FileSpec &file, uint64_t Size,
+  if (strtab == nullptr || strsz == nullptr)
+    return std::nullopt;
+
+  DataBufferSP data_sp = ReadMemory(process_sp, strtab->d_ptr, strsz->d_val);
```
The code above this checks that the ELF file was in memory, so this was only designed to work for in memory ELF files. But as you say we might be able to make this work for on disk files that have no section headers as well. Can we do this in a separate patch? I would like to make forward progress on ELF core files without making up new work.

clayborg committed:
Respond to user feedback.

clayborg committed:
Fix comment.

clayborg committed:
Respond to user feedback.

- remove changes the Process.h and Process.cpp
- Modify ObjectFileELF::CreateMemoryInstance(...) to read at least enough bytes to get the program headers.
- Modify ObjectFileELF::GetImageInfoAddress(...) to re-use the resolved address of the current dynanamic entry.
- Change ObjectFileELF::GetDynstrData(...) to use DT_STRTAB and DT_STRSZ even on ELF files with no section headers
- Add a new test to test .dynamic is found and parsed for files without section headers and not loading from memory.
- Update existing test comments to reflect what the test is doing.

clayborg committed:
Update comments to reflect the current approach to the GetDynamicData() and GetDynstrData() functions.

labath commented on lldb/test/Shell/ObjectFile/ELF/elf-dynamic-no-shdrs.test:
```diff
@@ -0,0 +1,49 @@
+// REQUIRES: system-linux, native
```
This test still requires linux a bunch of system dependencies. Here's a self-contained version that can be run anywhere (just run the input through yaml2obj and give it to lldb). I took `test/tools/llvm-readobj/ELF/needed-libs.test:` and adapted it for our use case -- I added DT_DEBUG as that's important for lldb, but you can also easily add any other tag you deem necessary
```
--- !ELF
FileHeader:
  Class:   ELFCLASS64
  Data:    ELFDATA2LSB
  Type:    ET_EXEC
  Machine: EM_X86_64
Sections:
  - Type:      SectionHeaderTable
    NoHeaders: true
  - Name:  .dynstr
    Type:  SHT_STRTAB
    Flags: [ SHF_ALLOC ]
    Content: '00616161006262620063636300' ## 0,a,a,a,0,b,b,b,0,c,c,c,0
    Size:    0x100
  - Name:    .dynamic
    Type:    SHT_DYNAMIC
    Address: 0x100
    Entries:
      - Tag:   DT_STRTAB
        Value: 0x0000000000000000
      - Tag:   DT_NEEDED
        Value: 9
      - Tag:   DT_NEEDED
        Value: 1
      - Tag:   DT_NEEDED
        Value: 5
      - Tag:   DT_STRSZ
        Value: 0x100
      - Tag:   DT_DEBUG
        Value: 0xdeadbeef
      - Tag:   DT_NULL
        Value: 0x0
ProgramHeaders:
  - Type: PT_LOAD
    VAddr: 0x0
    FirstSec: .dynstr
    LastSec:  .dynamic
  - Type:            PT_DYNAMIC
    FirstSec:        .dynamic
    LastSec:         .dynamic
    VAddr:           0x100
    Align:           0x8
```

clayborg committed:
Address user feedback.

- Fix indentation
- Call addr.GetSection()->GetFileAddress()
- Switch test to yaml

labath reviewed (approved):
Thanks for your patience.

clayborg committed:
run clang-format.

adrian-prantl commented:
@clayborg Based on the blamelist in https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/9352/ this broke [lldb-api.functionalities/gdb_remote_client.TestGdbClientModuleLoad.py](https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/9352/testReport/junit/lldb-api/functionalities_gdb_remote_client/TestGdbClientModuleLoad_py/)
Can you investigate/revert?

thanks!

clayborg commented:
I am on it!

clayborg commented:
I have identified the issue: with the new changes we are not successfully getting the address of the DT_DEBUG when calling:
```
Address ObjectFileELF::GetImageInfoAddress(Target *target);
```
Working on a fix.

clayborg commented:
I found the fix. Will have it in in a few minutes.
